### PR TITLE
Fixed #33004 -- Made saving objects with unsaved GenericForeignKey raise ValueError.

### PR DESCRIPTION
**Pull request by sarahboyce**
Tracker ticket: https://code.djangoproject.com/ticket/33004
GFK fields would raise an IntegrityError on save/bulk_create when they have an unsaved related model
Changed to a ValueError to allign with the behaviour of OneToOneField and ForeignKey fields.

Thank you @jonnythebard  for the initial patch: https://github.com/django/django/pull/14807 :muscle: 

Tried to address the review comments

### CODE REVIEW

File: tests/generic_relations/tests.py [hunk 1]
```diff
@@ -501,14 +500,36 @@ def test_query_content_type(self):
...
+        quartz = Mineral(name="Quartz", hardness=7)
+        tagged_item = TaggedItem(tag="shiny", content_object=quartz)
+        msg = (
+            "bulk_create() prohibited to prevent data loss due to unsaved related "
+            "object 'content_object'."
+        )
+        with self.assertRaisesMessage(ValueError, msg):
+            TaggedItem.objects.bulk_create([tagged_item])
+
+    def test_save_unsaved_instance_on_generic_foreign_key(self):
```
Comment (felixxm) [inline tests/generic_relations/tests.py:525]:
```suggestion
    def test_save_generic_foreign_key_after_parent(self):
```

File: django/db/models/base.py [hunk 2]
```diff
@@ -1108,6 +1108,26 @@ def _prepare_related_fields_for_save(self, operation_name, fields=None):
...
+                    continue
+                if obj.pk is None:
+                    raise ValueError(
+                        "%s() prohibited to prevent data loss due to unsaved "
+                        "related object '%s'." % (operation_name, field.name)
+                    )
+                elif getattr(self, field.fk_field) is None:
+                    # Use pk from related object if it has been saved after
+                    # an assignment.
+                    setattr(self, field.fk_field, obj.pk)
```
Comment (felixxm) [inline django/db/models/base.py:1129]: This is a separate issue and we should move it to a separate commit (together with `test_save_generic_foreign_key_after_parent`) e.g.
> _"Refs #28147 -- Fixed loss of assigned parent when saving GenericForeignKey after parent."_
Maybe even a separate PR to make it easier for review.

File: tests/generic_relations_regress/tests.py [hunk 3]
```diff
@@ -140,14 +139,37 @@ def count_places(place):
...
+            Note.objects.bulk_create([note])
+
+    def test_save_target_model_is_unsaved(self):
+        dev1 = Developer(name="Joe")
+        note = Note(note="Deserves promotion", content_object=dev1)
+        dev1.save()
+        # ValueError is not raised as unsaved object since been saved.
+        note.save()
+        note.refresh_from_db()
+        self.assertEqual(note.content_object, dev1)
```
Comment (felixxm) [inline tests/generic_relations_regress/tests.py:171]: It looks like a duplicate of `test_save_unsaved_instance_on_generic_foreign_key` :thinking:

File: django/db/models/base.py (hunk 2 above)
Comment (felixxm) [reply django/db/models/base.py:1129]: We should also check if this doesn't introduce a regression as 519016e5f25d7c0a040015724f9920581551cab0 (fixed in 4122d9d3f1983eea612f236e941d937bd8589a0d).

File: django/db/models/base.py [hunk 4]
```diff
@@ -1108,6 +1108,26 @@ def _prepare_related_fields_for_save(self, operation_name, fields=None):
...
+            if (
+                field.is_relation
+                and field.is_cached(self)
+                and hasattr(field, "fk_field")
+            ):
+                obj = field.get_cached_value(self, default=None)
+                if not obj:
+                    continue
+                if obj.pk is None:
+                    raise ValueError(
```
Comment (felixxm) [inline django/db/models/base.py:1122]: Should we remove the object from a cache? :thinking:

File: django/db/models/base.py [hunk 5]
```diff
@@ -1108,6 +1108,26 @@ def _prepare_related_fields_for_save(self, operation_name, fields=None):
...
+                        "%s() prohibited to prevent data loss due to unsaved "
+                        "related object '%s'." % (operation_name, field.name)
```
Comment (felixxm) [inline django/db/models/base.py:1124]: We can use a f-string
```suggestion
                        f"{operation_name}() prohibited to prevent data loss due to "
                        f"unsaved related object '{field.name}'."
```

**Review by felixxm — COMMENTED** [review]
@sarahboyce Thanks for this patch :+1: I left comments.

File: tests/generic_relations/tests.py (hunk 1 above)
Comment (sarahboyce) [reply tests/generic_relations/tests.py]: Thank you - save unsaved was a really bad name :joy:

File: django/db/models/base.py (hunk 2 above)
Comment (sarahboyce) [reply django/db/models/base.py]: I've removed and I'll raise a separate PR once we're happy with this one :+1:

File: tests/generic_relations_regress/tests.py (hunk 3 above)
Comment (sarahboyce) [reply tests/generic_relations_regress/tests.py]: Removed - I had the same feeling with the other two tests above (`test_target_model_is_unsaved_save` vs `test_unsaved_instance_on_generic_foreign_key_save` and `test_target_model_is_unsaved_bulk_create` vs `test_unsaved_instance_on_generic_foreign_key_bulk_create`) but as two tests existed before, I wasn't sure if I had missed something
Shall I also remove `test_target_model_is_unsaved_save` and `test_target_model_is_unsaved_bulk_create`?

Comment (felixxm) [reply tests/generic_relations_regress/tests.py]: > Shall I also remove `test_target_model_is_unsaved_save` and `test_target_model_is_unsaved_bulk_create`?

Yes, they are redundant.

File: django/db/models/base.py (hunk 4 above)
Comment (sarahboyce) [reply django/db/models/base.py:1122]: https://github.com/django/django/blob/6f311c7e35090957dd93783b71c5d3af7d5329d4/tests/one_to_one/tests.py#L153
In `OneToOneTest.test_unsaved_object` there's a check around it being removed from the cache
I'm not sure if it is needed in this case? I wouldn't know how to structure the test for the behaviour we want.
But it may be needed, I'm not sure.

File: tests/generic_relations_regress/tests.py (hunk 3 above)
Comment (sarahboyce) [reply tests/generic_relations_regress/tests.py]: They're dead :+1:

File: django/db/models/base.py [hunk 6]
```diff
@@ -1094,8 +1094,8 @@ def _prepare_related_fields_for_save(self, operation_name, fields=None):
                     if not field.remote_field.multiple:
                         field.remote_field.delete_cached_value(obj)
                     raise ValueError(
-                        "%s() prohibited to prevent data loss due to unsaved "
-                        "related object '%s'." % (operation_name, field.name)
+                        f"{operation_name}() prohibited to prevent data loss due to "
+                        f"unsaved related object '{field.name}'."
```
Comment (felixxm) [inline django/db/models/base.py:1098]: Please revert this, it's not worth changing existing code.

File: tests/generic_relations/tests.py [hunk 7]
```diff
@@ -501,14 +500,27 @@ def test_query_content_type(self):
         with self.assertRaisesMessage(FieldError, msg):
             TaggedItem.objects.get(content_object="")
 
-    def test_unsaved_instance_on_generic_foreign_key(self):
-        """
-        Assigning an unsaved object to GenericForeignKey should raise an
-        exception on model.save().
-        """
+    def test_unsaved_instance_on_generic_foreign_key_save(self):
```
Comment (felixxm) [inline tests/generic_relations/tests.py:503]:
```suggestion
    def test_unsaved_generic_foreign_key_parent_save(self):
```

File: tests/generic_relations/tests.py [hunk 8]
```diff
@@ -501,14 +500,27 @@ def test_query_content_type(self):
...
+        quartz = Mineral(name="Quartz", hardness=7)
+        tagged_item = TaggedItem(tag="shiny", content_object=quartz)
+        msg = (
+            "save() prohibited to prevent data loss due to unsaved related object "
+            "'content_object'."
+        )
+        with self.assertRaisesMessage(ValueError, msg):
+            tagged_item.save()
+
+    def test_unsaved_instance_on_generic_foreign_key_bulk_create(self):
```
Comment (felixxm) [inline tests/generic_relations/tests.py:514]:
```suggestion
    def test_unsaved_generic_foreign_key_parent_bulk_create(self):
```

File: tests/generic_relations/tests.py [hunk 9]
```diff
@@ -501,14 +500,27 @@ def test_query_content_type(self):
...
         quartz = Mineral(name="Quartz", hardness=7)
-        with self.assertRaises(IntegrityError):
-            TaggedItem.objects.create(tag="shiny", content_object=quartz)
+        tagged_item = TaggedItem(tag="shiny", content_object=quartz)
+        msg = (
+            "bulk_create() prohibited to prevent data loss due to unsaved related "
+            "object 'content_object'."
+        )
+        with self.assertRaisesMessage(ValueError, msg):
+            TaggedItem.objects.bulk_create([tagged_item])
```
Comment (felixxm) [inline tests/generic_relations/tests.py:523]: We could add a similar test for `bulk_update()`.

File: django/db/models/base.py (hunk 6 above)
Comment (sarahboyce) [reply django/db/models/base.py]: No worries, reverted

File: django/db/models/base.py (hunk 4 above)
Comment (felixxm) [reply django/db/models/base.py:1122]: As far as I'm aware it's unneeded, because a reverse relation `GenericRelation` returns multiple objects, so a single `tag` is not assigned to an `object`.

File: tests/generic_relations/tests.py [hunk 10]
```diff
@@ -501,14 +500,27 @@ def test_query_content_type(self):
         with self.assertRaisesMessage(FieldError, msg):
             TaggedItem.objects.get(content_object="")
 
-    def test_unsaved_instance_on_generic_foreign_key(self):
-        """
-        Assigning an unsaved object to GenericForeignKey should raise an
-        exception on model.save().
-        """
+    def test_unsaved_instance_on_generic_foreign_key_save(self):
+        # Assigning an unsaved object to a GFK should raise a ValueError on save.
```
Comment (felixxm) [inline tests/generic_relations/tests.py:504]: I'd chop these comments.

File: django/db/models/base.py [hunk 11]
```diff
@@ -1108,6 +1108,22 @@ def _prepare_related_fields_for_save(self, operation_name, fields=None):
                 ):
                     field.delete_cached_value(self)
 
+        # GenericForeignKey fields are private
```
Comment (felixxm) [inline django/db/models/base.py:1111]:
```suggestion
        # GenericForeignKeys are private.
```

**Review by felixxm — COMMENTED** [review]
@sarahboyce Thanks for updates :+1:

File: tests/generic_relations/tests.py (hunk 9 above)
Comment (sarahboyce) [reply tests/generic_relations/tests.py:523]: https://github.com/django/django/blob/6f311c7e35090957dd93783b71c5d3af7d5329d4/django/db/models/query.py#L736

We currently cannot bulk_update GenericForeignKeys as they are not a concrete fields (I could change this)

https://github.com/django/django/blob/6f311c7e35090957dd93783b71c5d3af7d5329d4/tests/queries/test_bulk_update.py#L166

I can explicitly add a GenericForeignKey case explicitly to this test if we want?

Comment (felixxm) [reply tests/generic_relations/tests.py:523]: Ahh true, sorry for the noise. No changes are required.

File: tests/generic_relations_regress/tests.py [hunk 12]
```diff
@@ -15,7 +14,6 @@
     Contact,
     Content,
     D,
-    Developer,
```
Comment (felixxm) [inline tests/generic_relations_regress/tests.py:18]: You can rebase from the `main` branch and remove `Developer` model, that is now unused (see also 1ed8ca43f61138b8f8d6f92106c27060753ed4e7)

File: tests/generic_relations/tests.py (hunk 10 above)
Comment (sarahboyce) [reply tests/generic_relations/tests.py]: Removed (chop means remove right?)

Comment (felixxm) [reply tests/generic_relations/tests.py]: Yes.

File: tests/generic_relations/tests.py (hunk 9 above)
Comment (sarahboyce) [reply tests/generic_relations/tests.py:521]: Ah I added something because I had caused a problem not handling the fields being specified in bulk_update - let me know if the test is useful or too much

File: tests/generic_relations_regress/tests.py (hunk 12 above)
Comment (sarahboyce) [reply tests/generic_relations_regress/tests.py:18]: Done :)

**Comment by felixxm** [comment]
@sarahboyce Thanks :+1:

File: tests/generic_relations/tests.py (hunk 9 above)
Comment (felixxm) [reply tests/generic_relations/tests.py:522]: I removed it.

**Comment by sarahboyce** [comment]
Thank you for the review @felixxm :tada:
